File: Sources/CountriesFlagsObjC/source/Countries/UIImage+Suriname.m
//
//  UIImage+Suriname.m
//  CountriesFlags
//

#import "Countries/UIImage+Suriname.h"

#import "Extensions/UIBezierPath+Star.h"

@implementation UIImage (Suriname)

+ (UIImage *)flagSurinameWithSize:(CGSize)size {
    UIColor* colorGreen = [UIColor colorWithRed:55.0f/255.0f
                                          green:126.0f/255.0f
                                           blue:63.0f/255.0f
                                          alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    UIColor* colorRed = [UIColor colorWithRed:180.0f/255.0f
                                        green:10.0f/255.0f
                                         blue:45.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorYellow = [UIColor colorWithRed:236.0f/255.0f
                                           green:200.0f/255.0f
                                            blue:29.0f/255.0f
                                           alpha:1.0f];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorGreen setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 20.0f * 4.0f, size.width, size.height / 20.0f * 12.0f));
    
    [colorRed setFill];
    UIRectFill(CGRectMake(0.0f, size.height / 20.0f * 6.0f, size.width, size.height / 20.0f * 8.0f));
    
    [colorYellow setFill];
    CGFloat radius = size.height / 20.0f * 8.0f / 2.0f;
    CGPoint starCenter = CGPointMake(size.width / 2.0f,
                                     (size.height + (size.height / 20.0 * 0.381966011 * 2.0)) / 2.0f);
    [[UIBezierPath starAtPoint:starCenter radius:radius type:FlagStarTypeFivePointed] fill];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+Suriname.swift
//
//  UIImage+Suriname.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Flag_of_Suriname "wikipedia"
     
     Generates the flag of Suriname.
     For more information see [wikipedia].
     
     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of Suriname.
     */
    internal static func flagSuriname(with size: CGSize) -> UIImage? {
        let colorGreen = UIColor(red: 55.0/255.0,
                                 green: 126.0/255.0,
                                 blue: 63.0/255.0,
                                 alpha: 1.0)
        let colorWhite = UIColor.white
        let colorRed = UIColor(red: 180.0/255.0,
                               green: 10.0/255.0,
                               blue: 45.0/255.0,
                               alpha: 1.0)
        let colorYellow = UIColor(red: 236.0/255.0,
                                  green: 200.0/255.0,
                                  blue: 29.0/255.0,
                                  alpha: 1.0)

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorGreen.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 20.0 * 4.0),
                          size: CGSize(width: size.width, height: size.height / 20.0 * 12.0)))
        
        colorRed.setFill()
        UIRectFill(CGRect(origin: CGPoint(x: 0.0, y: size.height / 20.0 * 6.0),
                          size: CGSize(width: size.width, height: size.height / 20.0 * 8.0)))
        
        colorYellow.setFill()
        let radius = size.height / 20.0 * 8.0 / 2.0
        UIBezierPath.star(with: CGSize(width: size.width, height: size.height + (size.height / 20.0 * 0.381966011 * 2.0)),
                          radius: radius).fill()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
